Sherlock/sherlock.h

```diff
@@ -139,111 +98,73 @@ template <typename ENTRY, template <typename> class PERSISTENCE_LAYER = DEFAULT_
 class StreamImpl {
  public:
   using entry_t = ENTRY;
-  using persistence_layer_t = PERSISTENCE_LAYER<ENTRY>;
-
-  // The inner structure containing all the data required for the stream to function.
-  // TODO(dkorolev): Make it a `ScopeOwnedByMe` type of thing.
-  struct StreamData {
-    persistence_layer_t persistence;
-
-    current::WaitableTerminateSignalBulkNotifier notifier;
-    std::mutex mutex;
-
-    template <typename... ARGS>
-    StreamData(ARGS&&... args)
-        : persistence(std::forward<ARGS>(args)...) {}
-  };
+  using persistence_layer_t = PERSISTENCE_LAYER<entry_t>;
+  using stream_data_t = StreamData<entry_t, PERSISTENCE_LAYER>;
 
   class StreamPublisher {
    public:
-    explicit StreamPublisher(std::shared_ptr<StreamData> data) : data_(data) {}
+    explicit StreamPublisher(ScopeOwned<stream_data_t>& data) : data_(data, []() {}) {}
 
     template <current::locks::MutexLockStatus MLS>
-    idxts_t DoPublish(const ENTRY& entry, const std::chrono::microseconds us = current::time::Now()) {
-      current::locks::SmartMutexLockGuard<MLS> lock(data_->mutex);
+    idxts_t DoPublish(const entry_t& entry, const std::chrono::microseconds us = current::time::Now()) {
+      if (!data_) {
+        CURRENT_THROW(StreamInGracefulShutdownMode());
```

Comment:
Btw, in Storage we have `StorageInGracefulShutdownException`. Let's define some common style for these exceptions :)